### ISSUE TITLE
Create a subgraph that indexes all the metadata for each DApp to replace mongoDB

### ISSUE THREAD

**Issue by andytudhope** [dap-ps/discover#42]
This is a fairly open bounty, ideally meant to be worked on as a hack at or around ETHBerlin.

The goal is clear though: use [The Graph](https://thegraph.com/docs/quick-start) to build a subgraph of all the [dapp metadata](https://github.com/dap-ps/discover/blob/master/back-end/models/dapps-metadata-model.js#L12) we're interested in that can be used to power the React app, rather than relying on MongoDB.

This will mean (at least):

1. Making sure you can use The Graph with Embark (and documenting those steps for both teams).
2. Running a graph node and building a subgraph of everything the React frontend needs.
3. Adapting the current code to handle slightly different queries/architecture.

## Notes

I promise to provide a 800 DAI through this bounty (via a tip) to a solution the solves all 3 steps above and provides a working solution that makes running mongoDB unnecessary. The bounty here is 1 DAI because I can't also pay Gitcoin 80 DAI for the pleasure. 

@bgits has apparently done some work like this, so perhaps can provide assistance? https://twitter.com/BarryGitarts/status/1161765205740793856

It may be best to do this in a separate repo for now, as Barry suggests in the above tweet. Either way, before any bounty work is started, you will need to write a short proposal for work in this issue. It may be OK to just work on a separate branch in this repo.

@yanivtal might want to track this issue too. As far as I understand, The Graph still runs their own Gateway and/or Query nodes, and so I'm not convinced it is 'decentralised' in practice, but it'd nevertheless be good to have ranked DApp data spread as far and wide as possible.

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 1.0 DAI (1.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/dap-ps/discover/42/3341).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/dap-ps/discover/42/3341).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $138,304.00 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 1 month from now.
Please review their action plans below:


**1) [iamonuwa](https://gitcoin.co/profile/iamonuwa) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/dap-ps/discover/42/3341?mutate_worker_action=approve&worker=iamonuwa) | [reject worker](https://gitcoin.co/issue/dap-ps/discover/42/3341?mutate_worker_action=reject&worker=iamonuwa))_.**

Let's do this. 😁. Subgraph it is...

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/dap-ps/discover/42/3341).

**Comment by andytudhope** [comment]
No @iamonuwa - I want you to work on the UI issues you have committed to. This is not a priority.

**Comment by andytudhope** [comment]
Reference repo: https://github.com/status-im/liquid-funding-subgraph

**Comment by nanspro** [comment]
@andytudhope i have worked with graph before and am familiar with it's architecture. Here is an sample [subgraph](https://github.com/nanspro/LocalEthereum) i created for localethereum. I see you need custom metadata as shown in `DAppsMetadataSchema` which means creating custom schema models, easy to do.
I guess you are currently using mongodb and your app reads data from that. After creating a subgraph you'll be to access that data faster and using graphql.
TBH i have never used embark so will have to dive into that. Saw Barry's comment and i think for completing this task frontend should also be modified to call proper queries.
Let me know what you think. I'll definitely not abandon this :)

**Comment by andytudhope** [comment]
Hmmm, @iamonuwa applied first, but your experience and that LocalEthereum example are really persuasive. I am happy for Isaac to work on all the UI issues, which will end up amounting to a similar amount in all likelihoood.

OK @nanspro - let's do this thing. If he finishes with UI work, perhaps @iamonuwa can help adapt the frontend to also get in on the bounty action here... Let's see how it goes. (Also, welcome back! It's great to see you again in another bounty :heart: )

**Comment by nanspro** [comment]
Awesome :smile: 
Starting on this in 9 hours!!

**Comment by hackmyway** [comment]
Can I also work on this project?
I recently completed a production app built on top of graph node with deployment of 2 subgraphs on it.

**Comment by andytudhope** [comment]
Hey @abhaykumar - thanks for the offer, but I don't have the funds (or review time!) to handle 2 separate implementations. How do you feel about web3 work in general though? I want to make how we handle transactions much more intelligent. 

Currently there is far too long a pause between a tx actually confirming, and it showing up as confirmed on the site. If we want to leverage GraphQL, we'll likely have to adapt the tx flow anyway, so it couldn't hurt to start improving it now, and make sure that it's as fast as it can be before we think about switching to a subgraph to really make this site ridiculously quick.

## Notes

1. I wouldn't depend on an interval to check if the transaction is mined or not: https://github.com/dap-ps/discover/blob/master/src/modules/TransactionStatus/TransactionStatus.reducer.js#L40

2. The [broadcastFunction](https://github.com/dap-ps/discover/blob/master/src/common/blockchain/services/helpers.js) should be able to dispatch an action once it receives a receipt:
```js
export const broadcastContractFn = (contractMethod, account, callback) => {
  return new Promise((resolve, reject) => {
    contractMethod
      .estimateGas({ from: account })
      .then(estimatedGas => {
        contractMethod
          .send({ from: account, gas: estimatedGas + 1000 })
          .on('transactionHash', hash => {
            resolve(hash)
          })
          .on('error', error => {
            reject(error)
          })
          .on('receipt', receipt => {
            // TODO: Dispatch an action or something
            callback(receipt);
          })
      })
      .catch(error => reject)
  })
}
```
then, wherever broadcastContractFn is called, pass as the last parameter what's supposed to happen once the transaction confirms. That action would receive a receipt, and do something with it (probably set some state in redux)...

**Comment by andytudhope** [comment]
@nanspro - what's the progress been like here? Have been away from a while, hence the delay in following up, but it would be good to see some work soon, otherwise my DAI needs to go and work in Compound instead of waiting to be transferred to you :wink:

**Comment by nanspro** [comment]
@andytudhope  understood the full backend code and started creating schema. I actually shifted my focus to this gitcoin hackathon recently :), will resume this and wrap it up by weekend. Save that DAI :sweat_smile:

**Comment by andytudhope** [comment]
Haha, ok cool. Thanks for the update, and I am happy to hear that you're part of the hackathon - big supporter of that. Not for this one as I am still catching up, but it would be fun to work together in another hackathon, so keep that in mind for the next one :wink: Will hold onto the DAI and await a PR here at your leisure :+1:

**Comment by gitcoinbot** [comment]
@nanspro Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=1) | [3 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=3) | [5 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=5) | [10 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=10) | [100 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=100)

**Comment by nanspro** [comment]
Hey @andytudhope  just 2-3days max. The hackathon deadline is tomorrow and I am working on some good stuff 😅

**Comment by gitcoinbot** [comment]
@nanspro Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [x] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=1) | [3 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=3) | [5 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=5) | [10 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=10) | [100 days](https://gitcoin.co/issue/dap-ps/discover/42/3341?snooze=100)

**Comment by andytudhope** [comment]
Hey @nanspro - a little tired of waiting here now - when this is escalated to mods, I'm gonna have to hand the issue and bounty off to someone else...

**Comment by nanspro** [comment]
Hey @andytudhope  i was working on this from last night and even talked to @yanivtal regarding some genuine issues which i encountered. Just now i was told a temporary fix by graph team regarding a weird issue that occurred. I plan to finish it today.

I completely understand the reason of removing me and i apologize for holding it for so long. However i think since i am very familiar with it by now i can finish it today and sooner than anyone who starts it now. Again really sorry for holding it for so long, definitely need to improve my time management skills.

P.S I didn't reply earlier because i thought that it'll escalate to mods by 9:45 PM tonight and i planned to finish it before that (7 hours left) and i think i'll be able to do that. You can check out my progress here https://github.com/nanspro/discover

**Comment by nanspro** [comment]
Hey @andytudhope 
There seems to be some issue getting ipfsHash from the data on graph. Currently the graph is deployed here https://thegraph.com/explorer/subgraph/nanspro/discover and it fetches ipfsHash correctly. Ex: https://cloudflare-ipfs.com/ipfs/QmWy2n7ysdMdCBDkBgBGT1TYrmB99eq5AxeJAzw8UDr5WG But when i deploy on graph it gives `Timeout Error`. I am talking to the graph team for any workaround and will update you. 

```
@nanspro Unfortunately, files don't propagate well through IPFS and are not guaranteed to be available on a specific IPFS node. IPFS timeouts mean that the file can't be retrieved from the network. Until that is solved with something like Filecoin, you'll have to handle ipfs.cat returning null gracefully. 

If you know which files are needed, you can manually add and pin them to our IPFS node at https://api.thegraph.com/ipfs/.
```

**Comment by iamonuwa** [comment]
@nanspro have you tried increasing the timeout?

**Comment by iamonuwa** [comment]
I can help you with this

**Comment by nanspro** [comment]
@iamonuwa pinging on you discord.

**Comment by nanspro** [comment]
@andytudhope  quoting Jannis from graph:
```
@nanspro Do you know which IPFS files the contracts link to and which IPFS node they are stored on.

IPFS today is a bit like everyone running their own, isolated file server. Just because something is added to IPFS doesn't mean it's available everywhere. In most cases, the files are only available on the IPFS node they were uploaded to, because there are no incentives for others to keep files around.

If you know which IPFS files you need and where they are stored, we can add and pin them to our node.
```

So could you tell me the ipfs node url where these files are stored as of now and then graph team will sync it to their node.
P.S I know that i have lost this bounty but still want to complete this ASAP.

**Comment by andytudhope** [comment]
Hey everyone. Now it is my turn to apologise for delays. Have been on a different kind of trip lately. I am happy to split the bounty however is appropriate, as I deeply appreciate these kind of issues (though I think communication around them is critical to prevent this kind of thing).

We simply use Infura as our IPFS provider, so I can't tell you exactly which node they are stored on. However, you can easily fetch all the hashes by looking at the DAppCreated events here: https://analytics.status.im/dapps/

Should be enough for The Graph team to add and pin them wherever is easiest. I am travelling to Tokyo, so will still be tough for me to communicate/review, but lmk if this helps and if we can move forward here before DevCon?

**Comment by nanspro** [comment]
Hey @andytudhope lol no need to apologise,i should've done it earlier. Ohh and i'll also be leaving for devcon on 6th oct, i got an scholarship to attend :) 

As of now i am on it and will definitely try to create the subgraph atleast as soon as possible(before leaving for osaka), the issue was that i have the hashes but the graph node was not able to find them. Let me try to resolve it with them.

**Comment by bgits** [comment]
I had a similar issue, I ended up pinning directly to the graph IPFS node. I also had the graph index the ipfs data so even if content disappeared from IPFS it would be retrievable from the sub graph.

**Comment by nanspro** [comment]
Yes @bgits so i guess i need the node url from status so that i can pin the files present there, right? I have all the hashes but the graph is not able to locate them. Should i provide them all the hashes separately so that they pin them to their node?

**Comment by bgits** [comment]
You need the node url from the graph. Also, what is the data on ipfs, if it’s just a json, I would have the graph index that as well.

**Comment by nanspro** [comment]
@bgits yes it's just a json. Ex: 
```json
{"name":"Status Kudos","url":"https://kudos.statusnet.eth","description":"Send real value to your friends and colleagues and thank them in public.","category":"OTHER","image":"https://prod.dap.ps/metadata/image/QmPwakhJ7XGFE4UhANgSjaw6bx4jAbCYJi8tgcbei9ehqL","dateAdded":1565943398078,"uploader":"0x5f3371793285920351344a1eaaaa48d45e600652"}
```

**Comment by andytudhope** [comment]
@nanspro - after our too-rushed chat in a DevCon elevator, I was expecting to eventually see some progress here? Are you still caught up in other work?

**Comment by nanspro** [comment]
Hey @andytudhope no i am not caught up with any other work besides the ongoing chainlink hackathon. I expected to do this after coming back from japan on 18th but i directly went into withdrawal (it was my first intl. trip) and then did nothing for 4-5 days.

But i am feeling pretty good and more motivated now :smile:  so let's wrap it up asap. I have created the subgraph, code can be found here https://github.com/nanspro/discover and the deployed subgraph is https://thegraph.com/explorer/subgraph/nanspro/discover. Could you check and confirm if it looks good or not? 
Btw only i got 11 ipfsHashes from `DappCreated` events on mainnet contract so does that mean that there are only 11 dapps data we need?
I am moving to the backend integration now. Let me know if anything needs to be added to the graph(more hashes or stuff).
Also i didn't understand how to add email to this data. Email is not present on ipfs so i guess that is something backend will have to handle on itself rather than relying on graph.

**Comment by andytudhope** [comment]
Hi @nanspro - welcome back to the land of the living :wink: Happy to hear you are recovered after your first trip.

There are only 11 DApps on mainnet atm, so that is correct. Also, the email stuff is correct as we just request it to contact the teams on submission, we don't store it publicly and don't need it after sorting out 1 or 2 issues with submissions, so you're correct that it's only our backend that needs to handle it. 

However, I'm not sure that we need to integrate the graph into our backend. I want the data on The Graph and I want to make sure that it continues getting added to so that it genuinely is easy to fork this and implement your own rules/token if you like. However, I don't want to use The Graph for our backend.

I am actually gonna put up a bounty to switch over to https://subspace.status.im because I think it is cooler than using any kind of backend. So, if you're happy that any other DApp which is added will appear correctly on The Graph and that our data will continue to be "backed up" there and easily available to anyone, then I think we can consider this bounty done...

**Comment by bgits** [comment]
Just happen to see this in my feed so chiming in. The Graph shouldn't need to be used by a backend, it should ideally be referenced in the front-end allowing each client to connect to the graph network separately. 

The Graph currently provides a few things that are hard to achieve otherwise.

- No requirement for a web3 provider. Even standard browsers can connect and at least view the data.
- [A clear roadmap to a performant incentivized read layer](https://thegraph.com/blog/the-graph-network-in-depth-part-1). [Subspace requires a web3 provider](https://subspace.status.im/getting-started.html#connecting-to-a-web3-provider)  which means most users of the dap.ps will just be using Infura (which AFAIK still has no plan for incentivization). Relying only on the JSON-RPC can be slow to sync if the the dapp has generated a lot of data and/or error-prone as the JSON-RPC doesn't return standard errors (ie: 0 instead of an error) and can lead to data corruption of the client side db if the responses are not somehow validated.

Just some food for thought

**Comment by nanspro** [comment]
> I'm not sure that we need to integrate the graph into our backend. I want the data on The Graph and I want to make sure that it continues getting added to so that it genuinely is easy to fork this and implement your own rules/token if you like. However, I don't want to use The Graph for our backend.

Yeah my bad. The graph will remove the need for storing such stuff in backend so i was referring to changing backend code and adding graph to frontend. Yes whenever a new dapp is created, it will be accessible via graph.

Ohh one more thing. The graph works perfectly in getting ipfs hashes of content but out of 11 hashes, two don't work. These two ipfs hashes `QmS6a72GnPvUCMwKKrVGE41yY8RYwVVoBTrEbW6XWDu1EY` and `QmfCbEDwZ7sVSzcmivp3WvKd9pcKHhmCXiFwFuuQJmhPhs` are not accessible by my node and i asked the graoh team and we tried with 3-4 different gateways but can't access the content present here. Could you access these hashes and do you know why this might happen?

@andytudhope

**Comment by andytudhope** [comment]
Yeah, cool. Am still chewing on how best to combine subspace and The Graph. While I do that

`QmS6a72GnPvUCMwKKrVGE41yY8RYwVVoBTrEbW6XWDu1EY` is Clovers Network and

`QmfCbEDwZ7sVSzcmivp3WvKd9pcKHhmCXiFwFuuQJmhPhs` is BULVRD Drive

I do not know why they are not accessible through IPFS :disappointed: @bakasura980 do you have any idea why this might happen? Potentially we failed to pin them once they were submitted.

Interestingly, Clovers is APPROVED, and BULVRD is still NEW (as they don't comply with all our requirements). So it's not got to do with whether they're approved or not...

**Comment by andytudhope** [comment]
I think it is because Clovers submitted twice once without staking, once with a stake, and their original hash is `QmR5aVSZzyNjuS18kpXrzk3dhxW1ZWRaxaZ6TGMzWfeQ32`, which does resolve correctly.

And then maybe BULVRD is not accessible because it has not been APPROVED yet - do we only pin to IPFS on approval @bakasura980?

**Comment by bakasura980** [comment]
Hello @andytudhope, I did not get any notifications, just enter here to check what is the status. As I remember, we did not have IPFS integrated when I stopped working on this, but yes, check out if you pin these records, because IPFS will `garbage collect` unvisited records.

**Comment by nanspro** [comment]
@bakasura980 i noticed just now that the subgraph is failing due to `Failed ipfs.cat, returning null`, this error indicates that the graph node is not picking up the new ipfs hashes and when earlier this error occurred, i synced with graph node using `ipfs-sync` to https://api.thegraph.com/ipfs/

Can't we just pin the node url to graph node so that the sync happens automatically?

**Comment by bakasura980** [comment]
@nanspro Try it, I am not familiar with subgraphs. An idea is to have a cron job which process syncing in every 24 hours for example

**Comment by nanspro** [comment]
@bakasura980 if you have guys have a public url of node which stores all the ipfs hashes then i can just pin that to graph's node and it'll keep listening for new files and hashes?

**Comment by andytudhope** [comment]
Yeah, there's nothing like that, though I am very confused by @bakasura980's comment, because we definitely do use IPFS whether he remembers coding it or not :sweat: I am also not familiar enough with The Graph to be helpful with the `null` error when trying to pin, though clearly there are at least 2 DApps which did not get correctly pinned in IPFS when they were submitted and hence are causing us issues. I'm honestly not sure how to go about fixing that, other than manually excluding the ones above I have already listed?

**Comment by andytudhope** [comment]
Hey @nanspro - I am going to close the bounty here for now seeing as we haven't been able to solve it, and I want to include this in the scope of work for v2 of the site in general. If you send me an email with your address, I will make sure you receive at least some compensation for the work done.
Leaving the issue open though, for reference.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Cancelled** 

<hr>

__The funding of 1.0 SAI  (1.0 USD @ $1.0/SAI) attached to this issue has been **cancelled** by the bounty submitter__

 
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $80,121.53 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by nanspro** [comment]
Ahh don't worry about the compensation @andytudhope  :smile: 

This error is really weird i suppose anyone from graph team can maybe help with this. I hope the subgraph i created can be used as a reference.

**Comment by andytudhope** [comment]
Hey @nanspro 

Any chance you could have another go at this subgraph using our own IPFS stuff, rather than relying on a dodgy Infura connection? See https://github.com/dap-ps/discover/commit/9dee93c0773793b8222464b3055530989285aa0c for more detail on changes made.

**Comment by nanspro** [comment]
@andytudhope sure will have look in the morning! :+1:

**Comment by nanspro** [comment]
@andytudhope could you help me with accessing the new data from ipfs hashes. Subgraph founds new hashes but i can't view the content on ipfs.status.im/{hash} for hashes like `QmfMx6kvs4bandCC76NLUgWJ63PaA5GyZ3w1PzpXxE1PVc` , `QmfCbEDwZ7sVSzcmivp3WvKd9pcKHhmCXiFwFuuQJmhPhs`. 

I have also asked the graph team to pin ipfs.status.im:443 to their node so that we can resolve ipfs hashes but that will work only once data can be accessed on ipfs.status.im/{hash} first, which idk why isn't happening cause i can see you are pinning to the node in the backend :thinking:

**Comment by andytudhope** [comment]
Hey @jakubgs is this data accessible from our cluster? Or will you need to wave your magic wand to make it so?

**Comment by jakubgs** [comment]
Ah, you're right. I did pin all the images but I forgot that Dapp metadata is also stored on IPFS as JSON. I will have to go and fetch the JSON data in the right format and upload it to our IPFS cluster.

I'll try doing that this week. Here's the issue: https://github.com/status-im/infra-ipfs/issues/4

**Comment by jakubgs** [comment]
In https://github.com/status-im/infra-ipfs/issues/4 I managed to update the hashes to proper ones and uploaded the matadata to IPFS for https://dev.dap.ps/ but that broke "Highest Ranked" dapps view because all have `sntValue` of `0`, except one I added after that.

Maybe someone can help me figure out what is going on.

**Comment by jakubgs** [comment]
In https://github.com/status-im/infra-ipfs/issues/4 and https://github.com/dap-ps/discover/pull/76 I've made sure that `ipfaHash` is always set to the correct IPFS content hash. The `hash` field will contain and incorrect hash for older Dapps because of legacy reasons due to how the contract identifies the Dapps.

**Comment by andytudhope** [comment]
Hey @nanspro - can you try again, using the `ipfsHash` field, which should be correct, up-to-date and accessible for all the DApps now.

**Comment by andytudhope** [comment]
@nanspro let's work here: https://github.com/dap-ps/graph

Basically, when `hash == ipfsHash` we're all good. When `hash != ipfsHash`, then it is a legacy DApp and we need to fetch it's info using the `ipfsHash`, even though what is coming from the contract will match what is in the `hash` field not the `ipfsHash` one. I'm not actually sure how to get around this, as there is (in my mind) no good way to get the correct hash for legacy DApps based just on the info in the contract. But maybe I am wrong... This is what the schema for them now looks like:

```
{
    "_id" : ObjectId("5d555a235f0209249ce25f1c"),
    "status" : "APPROVED",
    "details" : {
        "name" : "Livepeer",
        "url" : "https://explorer.livepeer.org/",
        "description" : "Decentralized video broadcasting",
        "category" : "UTILITIES",
        "image" : "/metadata/image/QmTuCSBso1n9gqRhCVsnPJdqeR4jzvHaAen2Jk81kqJXnJ",
        "dateAdded" : 1565874722808.0,
        "uploader" : "0x1d32f2acb832afc3d8c8ffb3be20e8dc7faac507"
    },
    "compressedMetadata" : "0x99ce46406e4a1998c185bc114acaf3870055057404744f234c094c8091231a99",
    "email" : "andy@status.im",
    "hash" : "QmRFdYmKphoKHh1TocKdqDiY8D1RomRx3mS55fhwY1jM2H",
    "__v" : 0,
    "ipfsHash" : "QmRCXSJLNfJpWzC7hZBQoLMCdbzuSrs8PNR3sJsKREiRSV"
}
```

If you can help me get it working, this bounty is still yours :wink:

**Comment by jakubgs** [comment]
Well, the service would have to call https://dap.ps/metadata/all and find the DApp based on the key which is the legacy hash, and take the correct hash from `ipfsHash`, like so:
```bash
 $ curl -s https://raw.prod.dap.ps/metadata/all \
    | jq -r '.["QmPUEoCQXYB7k6sWsyecKn2vgCqYmtatnuVAjtuzHRi9Zt"].ipfsHash'
QmUjKQC9yWAN4YWyLSYWMS7TbAeRLUcmE5eCU8wX5vfsg
```

**Comment by andytudhope** [comment]
The primary goal here is to get the info out the contract and nowhere else, so that it really is decentralised and does not rely on anyone's service for correctness. However, I don't see that being a possibility, unless every DApp were to update their metadata themselves, which is not really an option at this time.

**Comment by jakubgs** [comment]
Or you can hardcode the legacy hashes into the data collecting software...

**Comment by nanspro** [comment]
@andytudhope i'll take a look at it again :+1:

**Comment by nanspro** [comment]
Sorry guys got really caught up with mid-sems and lot of work from my internship. I'll get time on this weekend, so will look into this.

**Comment by nanspro** [comment]
Okay so a lot happened here. Lemme make sure i get this right, because the image urls were changed the ipfs hashes were changed but from the contract i can only fetch the old ipfsHash so i have two options:
1. Get new hashes from the `ipfsHash` field by requesting this url https://dap.ps/metadata/all. I def don't like this way as it's a very centralized way of fetching data and defeats the whole purpose of using graph. Plus idk exactly how i'll make api calls from graph as it is heavily type constrained (it is assembly script actually instead of just typescript) But i'll ask the graph team about it.

2. Write if statements for legacy hashes and fetch data from new ipfsHash which is hardcoded whenever the app encounters a legacy hash. This looks good and make sense.

There seems to be no better way to do this as i can't get new ipfs hashes from contract so i guess we should go with 2nd approach. @andytudhope  @jakubgs  what do you guys think?

**Comment by jakubgs** [comment]
No, you don't need the 2nd step. The way it work is that `hash` contains either the old hash and `ipfsHash` contains the new hash after. The other case is that the dapps that were created after the fix will have both `hash` and `ipfsHash` set to the same value.

The contracts contain the old hashes for dapps created before fix and new correct hashes for dapps created after that. The reason why the hashes were incorrect was because the metadata was being uploaded as JSON, but __quotes__ in double quotes, which is stupid beyond belief.

**Comment by nanspro** [comment]
@jakubgs so the hashes that were incorrect, i'll have to use their ipfsHash to obtain the data from, right? I can see that a lot of hashes are incorrect (10-12 i think or more) so the only way i see is hardcoding their new values in the graph so that once the graph came across a hash which is incorrect, it'll use the new value.
Like this 
```js
 if(ipfsHash == 'QmS6a72GnPvUCMwKKrVGE41yY8RYwVVoBTrEbW6XWDu1EY')
    ipfsHash = 'QmaY9NPShDqWnGuBH9kCmmsvSkizRBqT5FMiZnRKxVQ1hR';
```
Right now i am manually checking which hashes are wrong, do you have a list or something of the incorrect hashes?

**Comment by nanspro** [comment]
Like this isn't working http://ipfs.status.im/QmfTABxDjZNrgeoGLuCSK44cNEWFmXPaKh2iBMPoMBqgPa but works with new hash http://ipfs.status.im/Qmc7pVDcu2GpgGsHfCxwDPYjBYvmCYK9DVwxpnctTs4A12

**Comment by nanspro** [comment]
@andytudhope 
> I'm not actually sure how to get around this, as there is (in my mind) no good way to get the correct hash for legacy DApps based just on the info in the contract.

I can't think of any way either except replacing incorrect hashes with the correct ones manually.
Also i'll have to manually sync these hashes with graph node as of now because as graph team told me:

> We're close to rolling out a feature that allows graph-node to check more than one IPFS node for files (in the case of the hosted service, more than just ours at https://api.thegraph.com/ipfs/). Unfortunately, all these gateways like Cloudflare, https://ipfs.status.im/ etc. don't serve an API that is compatible with IPFS nodes.

> We wrote a little tool that makes it easy to sync a list of files (note the --file-list argument) from one IPFS node to another: https://github.com/graphprotocol/ipfs-sync#usage (our node is at https://api.thegraph.com/ipfs/)

So in the future we'll have to regularly sync status's node with graph node. It's just a script but still without syncing it won't work(graph won't be able to fetch the data from ipfs)

**Comment by nanspro** [comment]
@jakubgs @andytudhope any updates on this? I don't think there's much to left, i am available to wrap it up whenever you are free.  But i am still unclear on some things so can't go ahead on my own :disappointed:

**Comment by jakubgs** [comment]
>Like this isn't working

Yes, old hashes were never properly pinned. Also the data was uploaded incorrectly, which is why the content hash has changed.

>I can't think of any way either except replacing incorrect hashes with the correct ones manually.

You mean by hardcoding a map in the service? If it's dumb and it works it's not dumb.

>So in the future we'll have to regularly sync status's node with graph node.

That's not an issue, I can do that. I assume if we just run a daily script to sync it with TheGraph cluster it would be enough?

**Comment by andytudhope** [comment]
Agree with the "If it's dumb and it works it's not dumb" comment. jakubs gist should help with creating a hardcoded map: https://gist.github.com/jakubgs/82ec4dde8687a565b1a5c16341c28660

Daily script FTW as far as I am concerned.

**Comment by nanspro** [comment]
@jakubgs @andytudhope 
> If it's dumb and it works it's not dumb

yeah i am gonna hardcode the hash values and then sync all hashes with graph node, that should work!!

**Comment by jakubgs** [comment]
Working on this in https://github.com/status-im/infra-ipfs/issues/5 but it doesn't seem to work so far.

**Comment by jakubgs** [comment]
It should be synced.

**Comment by nanspro** [comment]
@jakubgs unable to open the logs for my subgraph (https://thegraph.com/explorer/subgraph/nanspro/discover?version=current) talking to graph team about this, also there are so many old hashes present here https://dap.ps/metadata/all so i am putting if statements for them but there are too many lol

**Comment by jakubgs** [comment]
Why would you do if statements? Don't you know what a map is?

**Comment by nanspro** [comment]
ohh i totally forgot (sorry not used to typescript). Yeah will do map

**Comment by nanspro** [comment]
@jakubgs i can see around 100 hashes with new ipfshashes, do you have any idea on replacing them rather than me putting each old hash from https://dap.ps/metadata/all and setting it's new value to ipfsHash. I can do that but just wondering if there's any better way?

**Comment by jakubgs** [comment]
Sounds about right:
```
 > curl -s https://dap.ps/metadata/all \
    | jq '[.[] | { key: .hash, value: .ipfsHash } | select(.key != .value)] | from_entries' \
    | head
{                                                                                                                                            
  "QmSWxFAHpjRiydEBZdeQfzWmpLuc3YHBTnvYSkxZyM6Bj5": "QmZJnjgbH8z5QbLzNcPaCK4bMzvye5MpPTS4vSpB55XBfm",
  "QmNgNAvHccH7jAbvpCZFLKGNJ8pCZm7Jepmibg3tCSCBPC": "QmVoryjXtawJVRjELzRfmpXgGkzR6U2RJvXLjMXnDWUQrT",
  "QmRFdYmKphoKHh1TocKdqDiY8D1RomRx3mS55fhwY1jM2H": "QmRCXSJLNfJpWzC7hZBQoLMCdbzuSrs8PNR3sJsKREiRSV",
  "QmWXKBHcWqwyTM7hxWXSqVcTV2FVdb6NstsuUb9UVSjrWZ": "QmQFMrVpHweydbCYXwHq835egvjyTJDGJwMjgsopCJaHvz",
  "QmVmY5K5KffCCzNmMsZp7iDefNPjtQomh4eGERso7ShSFk": "Qme7uzZ7V66n93hKJWWaWs3RhH5oHWJqPtsS8DrvGRZZu6",
  "QmNxiRWB2pEt8PxZcqSk31rrf1Q2wJ6bwNqpgCNc5Yofnx": "QmYySAayyAGBheZR5bwZ6EnbhpZWLiTyym5Hziba8d2LTf",
  "QmQ8WdvGhByH3EeTLGVgApJMNjs8TD3Qb9wonVbtU9EYD6": "QmSNAtyoGsYL6tjV6STp3UJpsAd27CLpePxLjpBqR5AThy",
  "QmSrNSRfvAiP2WFkEjsW1zz3GqSGiAz9Ntpz8EqzHiiFRq": "QmSeuh6Q75iapu267SSvv1HRHRVRvymxfjCCtEhGWYzL7Y",
  "QmVYNinXCBt1sp44gP6HcAWKm9A8x2cb8zgenpGnjEorEo": "QmXEp49VDGUE8eZ7w8yieLTjZkZxeibuMQhw6EyYnLxwxv",
```
```
 > curl -s https://dap.ps/metadata/all \
    | jq '[.[] | { key: .hash, value: .ipfsHash } | select(.key != .value)] | from_entries' \
    | wc -l
110
```

There is no better way, we can't replace them in the DB because they have to match the old fucked up hashes in the contract. You can learn about why that is in https://github.com/status-im/infra-ipfs/issues/4.

**Comment by nanspro** [comment]
@jakubgs @andytudhope the subgraph is synced and working now. Will push my code to repo and create a PR now
